File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-1.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        ans = -1000000
        def dfs(root):
            
            nonlocal ans
            if not root:
                return 0
            
            left = max(0,dfs(root.left))
            right = max(0,dfs(root.right))

            curr = root.val + max(right, left, right + left)
            ans = max(ans, curr)
            # print(ans)
            return root.val + max(left, right)

        dfs(root)
        return ans

File: Data Structures & Algorithms/binary-tree-right-side-view/submission-1.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if(!root) return {};
        queue<TreeNode*> q;
        q.push(root);
        vector<int> ans;
        while(!q.empty()) {
            int n = q.size();
            int val = -101;
            for(int i = 0; i<n; i++){
                TreeNode* node = q.front();
                q.pop();
                val = node->val;
                if(node-> left != nullptr) q.push(node->left);
                if(node-> right != nullptr) q.push(node->right);
            }
            ans.push_back(val);
        }
        return ans;
    }
};
